refactor(compiler): implement factory declarations

petebacondarwin opened the pull request:
For public API reviewers, the only change is the removal of `ɵɵinjectPipeChangeDetectorRef` in  `goldens/public-api/core/core.d.ts`

petebacondarwin committed:
refactor(compiler-cli): implement `ɵɵngDeclareFactory`

This commit changes the partial compilation so that it outputs declaration
calls rather than compiled factory functions.

The JIT compiler and the linker are updated to be able to handle these
new declarations.

petebacondarwin committed:
refactor(compiler): remove `injectFn` option from factory metadata

The `injectFn` reference can be inferred unamiguously from the `target`
property so it is not needed.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/injectable.ts:
```diff
@@ -127,6 +121,43 @@ export class InjectableDecoratorHandler implements
     return results;
   }
+
+  compilePartial(node: ClassDeclaration, analysis: Readonly<InjectableHandlerData>):
```
For this PR, the `compilePartial()` is just a copy of the `compileFull()` with the exception that the factory is now declared rather than defined.

A follow up PR to implement `ngDeclareInjectable()` will tidy this up.

petebacondarwin replied:
Yep!

petebacondarwin commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+   */
+  target: R3FactoryTarget;
+
+  delegate?: o.Expression;
+  delegateType?: R3FactoryDelegateType;
+  delegateDeps?: R3DependencyMetadata[];
+
+  expression?: o.Expression;
```
It is possible that `ɵfac` factory declarations never actually use these properties.  They are definitely used when defining the `factory` property inside an Injectable definition, but that is different.  These properties may be a result of previous code sharing between `ɵfac` and `factory` implementations.

If this turns out to be the case, we can remove these and the associated logic in a future PR.

JoostK replied:
Yeah I think this is only used internally. I'd definitely vote to clean this up.

petebacondarwin replied:
I think we can remove it... Trying now.

petebacondarwin commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+  /**
+   * Injecting the `ChangeDetectorRef` token. Needs special handling when injected into a pipe.
+   */
+  ChangeDetectorRef = 2,
```
This value is slightly problematic as it is exposing a bit more of the underlying implementation details of how DI is wired up than we might like. There was a PR to allow us to remove this (https://github.com/angular/angular/pull/33072) but it was closed without being merged.

I looked into alternative methods for hiding this from the declaration, such as having a `CoreToken` value allowing us to simply check the name of the dependency (to see if it matches `ChangeDetectorRef`) internally but this does not work with JIT compilation, which itself relies upon a static property of the `ChangeDetectorRef` class.

petebacondarwin replied:
There was no way, that I could find, to remove the need to tell the injector if the factory is instantiating a pipe or not at runtime; but by moving this information into the `InjectorFlags` and passing that information through to the `bloomHash(flags)` special factory calls, we are able to remove this ``ChangeDetectorRef` target.

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_factory_linker_1.ts:
```diff
@@ -0,0 +1,100 @@
+      delegate: metaObj.getOpaque('delegate'),
+      delegateDeps: getDeps(metaObj, 'delegateDeps'),
+      delegateType: parseEnum(metaObj.getValue('delegateType'), R3FactoryDelegateType),
+    } as R3FactoryMetadata;
```
I don't think this cast adds anything as the return type is used as contextual type. Or maybe you meant to cast to `R3DelegatedFnOrClassMetadata`.

petebacondarwin replied:
Ideally I will try to remove these delegated types from the partial linker altogether. I don't think they are actually ever used...

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_factory_linker_1.ts:
```diff
@@ -0,0 +1,100 @@
+    };
+  }
+
+  return meta as R3FactoryMetadata;
```
Here as well.

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/util.ts:
```diff
@@ -7,7 +7,25 @@
+/**
+ * Parses the value of an enum from the AST value's symbol name.
+ */
+export function parseEnum<TExpression, TEnum>(
```
Neat!

JoostK commented on packages/compiler-cli/src/ngtsc/annotations/src/injectable.ts:
```diff
@@ -127,6 +121,43 @@ export class InjectableDecoratorHandler implements
+      results.push({name: 'ɵprov', initializer: res.expression, statements, type: res.type});
+    }
+
+
```
nit: too many newlines.

JoostK commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -419,7 +419,10 @@ function allTests(os: string) {
 
           const jsContents = env.getContents('test.js');
-          expect(jsContents).toContain('/** @pureOrBreakMyCode */ i0.ɵɵgetInheritedFactory(Dir)');
+          expect(jsContents).toContain('Dir.ɵfac = /** @pureOrBreakMyCode */ function () {');
```
I doubt the purity marker is still needed. There is no call expression any more, only a function expression.

petebacondarwin replied:
Actually this is now an IIFE, so there is a call expression. I think it is still valid to keep this comment.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+  name: string;
+
+  /** Number of arguments for the `type`. */
+  typeArgumentCount?: number;
```
No need for this in the partial declaration; this only exists to support creating the .d.ts type declaration.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+  /**
+   * String name of the type being generated (used to name the factory function).
+   */
+  name: string;
```
Why is this here? We use `type.getSymbolName()` in all places we need a typename.

petebacondarwin replied:
Good call.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+   */
+  target: R3FactoryTarget;
+
+  delegate?: o.Expression;
+  delegateType?: R3FactoryDelegateType;
+  delegateDeps?: R3DependencyMetadata[];
```
I have yet to decide if I like these as stable API.

petebacondarwin replied:
I think we should try to remove these properties altogether. I think they are no use for `fac` properties only for `factory` inside Injectable definitions.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+   * the unknown type if no literal type is available (e.g. the attribute name is an expression).
+   * Will be null otherwise.
+   */
+  attribute: o.Expression|null;
```
This is not needed in the partial declaration. This is captured in the compile metadata to be able to generate the attribute name into the .d.ts files if it's literal, but not otherwise used. If `R3ResolvedDependencyType` is `Attribute` then `token` will be the attribute expression.

Maybe, we should have `attribute` instead of `token`. Maybe that would even avoid the need for the `resolved` enum altogether, if we derive that from the partial declaration shape.

petebacondarwin replied:
I will remove this from this interface since it only refers to the type of an attribute, which the linker does not care about.
I think that we still need the `resolved` property though.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+  /**
+   * Whether the dependency has an @SkipSelf qualifier.
+   */
+  skipSelf: boolean;
```
I am not looking at the full diff so this may be inaccurate, but these should all be optional.

petebacondarwin replied:
Yep. All made optional.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+   * An enum indicating whether this dependency has special meaning to Angular and needs to be
+   * injected specially.
+   */
+  resolved: R3ResolvedDependencyType;
```
I'm not fond of the name `resolved` here. As for the enum name, the _Resolved_ name is implied by this being the partial declaration: _everything_ in the partial declaration has been resolved.

petebacondarwin replied:
So this was the original name before this PR.

We could rename it... but I think `type` would be confusing for a dependency since one can think of the `token` being the "type" of the dependency...

How about `depType`?

petebacondarwin replied:
OK, I'll give that a go.

petebacondarwin replied:
The union approach results in code like this in the linker:

```ts

function getDep<TExpression>(
    depObj: AstObject<R3DeclareTokenDependencyMetadata, TExpression>|
    AstObject<R3DeclareAttributeDependencyMetadata, TExpression>): R3DependencyMetadata {
  const options = getDepOptions(depObj);
  if (isTokenDep(depObj)) {
    return {...options, token: depObj.getOpaque('token')};
  } else if (isAttributeDep(depObj)) {
    return {...options, attribute: depObj.getOpaque('attribute')};
  } else {
    return {...options, token: o.literal(null)};
  }
}

function isTokenDep<TExpression>(depObj: AstObject<R3DeclareDependencyMetadata, TExpression>):
    depObj is AstObject<R3DeclareTokenDependencyMetadata, TExpression> {
  return (depObj as AstObject<R3DeclareTokenDependencyMetadata, TExpression>).has('token');
}

function isAttributeDep<TExpression>(depObj: AstObject<R3DeclareDependencyMetadata, TExpression>):
    depObj is AstObject<R3DeclareAttributeDependencyMetadata, TExpression> {
  return (depObj as AstObject<R3DeclareAttributeDependencyMetadata, TExpression>).has('attribute');
}

function getDepOptions<TExpression>(
    depObj: AstObject<R3DeclareDependencyMetadataOptions, TExpression>):
    R3DependencyMetadataOptions {
  return {
    host: depObj.has('host') && depObj.getBoolean('host'),
    optional: depObj.has('optional') && depObj.getBoolean('optional'),
    self: depObj.has('self') && depObj.getBoolean('self'),
    skipSelf: depObj.has('skipSelf') && depObj.getBoolean('skipSelf'),
  };
}
```

JoostK replied:
I would imagine it to look something like this?

```ts
function getDep<TExpression>(
    depObj: AstObject<R3DeclareTokenDependencyMetadata, TExpression>|
    AstObject<R3DeclareAttributeDependencyMetadata, TExpression>): R3DependencyMetadata {
  const options = getDepOptions(depObj);
  let token: o.Expression;
  let resolved: R3ResolvedDependencyType;
  if (isTokenDep(depObj)) {
    token = depObj.getOpaque('token');
    resolved = R3ResolvedDependencyType.Token;
  } else if (isAttributeDep(depObj)) {
    token = depObj.getOpaque('attribute');
    resolved = R3ResolvedDependencyType.Attribute;
  } else {
    throw new FatalLinkerError(...);
  } 

  return {
    token,
    resolved,
    host: depObj.has('host') && depObj.getBoolean('host'),
    optional: depObj.has('optional') && depObj.getBoolean('optional'),
    self: depObj.has('self') && depObj.getBoolean('self'),
    skipSelf: depObj.has('skipSelf') && depObj.getBoolean('skipSelf'),
  };
}

function isTokenDep<TExpression>(depObj: AstObject<R3DeclareTokenDependencyMetadata|R3DeclareAttributeDependencyMetadata, TExpression>):
    depObj is AstObject<R3DeclareTokenDependencyMetadata, TExpression> {
  return depObj.has('token');
}

function isAttributeDep<TExpression>(depObj: AstObject<R3DeclareTokenDependencyMetadata|R3DeclareAttributeDependencyMetadata, TExpression>):
    depObj is AstObject<R3DeclareAttributeDependencyMetadata, TExpression> {
  return depObj.has('attribute');
}
```

Not sure if that union type in the type guards works as I hope it would.

petebacondarwin replied:
I propose instead to go with a simpler single interface:

```ts
export interface R3DeclareDependencyMetadata {
  /**
   * An expression representing the token or value to be injected, or `null` if the dependency is
   * not valid.
   */
  token: o.Expression|null;

  /**
   * If this dependency is due to the `@Attribute()` decorator, then this is an expression
   * evaluating to the name of the attribute. Otherwise this property is not present.
   */
  attribute?: o.Expression;

  /**
   * Whether the dependency has an @Host qualifier.
   * Default: false,
   */
  host?: boolean;

  /**
   * Whether the dependency has an @Optional qualifier.
   * Default: false,
   */
  optional?: boolean;

  /**
   * Whether the dependency has an @Self qualifier.
   * Default: false,
   */
  self?: boolean;

  /**
   * Whether the dependency has an @SkipSelf qualifier.
   * Default: false,
   */
  skipSelf?: boolean;
}
```

petebacondarwin replied:
> I would imagine it to look something like this?
> ...
> Not sure if that union type in the type guards works as I hope it would.

Sadly TS is not able to split the `AstObject<T|U>` into `AstObject<T>` and `AstObject<U>` based on type guards 🤷‍♂️

petebacondarwin replied:
I went with the simpler interface. It nice and clean IMO.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,129 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+   * then one or more of the parameters wasn't resolvable and any attempt to use these deps will
+   * result in a runtime error.
+   */
+  deps: R3DependencyMetadata[]|'invalid'|null;
```
The `invalid` is interesting here. Individual deps can also be invalid. It's unclear to me when the whole thing is marked as invalid vs when the individual deps are marked as invalid. Maybe @alxhub knows.

petebacondarwin replied:
Here is the relevant PR: https://github.com/angular/angular/pull/33739
It is so that we can throw a more helpful error when a dependency is not valid.

JoostK reviewed (commented):
Did a semi-quick pass; have not at all looked into everything in detail.

petebacondarwin committed:
refactor(core): remove the need for `ɵɵinjectPipeChangeDetectorRef()`

This instruction was created to work around a problem with injecting a
`ChangeDetectorRef` into a pipe. See #31438. This fix required special
metadata for when the thing being injected was a `ChangeDetectorRef`.

Now this is handled by adding a flag `InjectorFlags.ForPipe` to the
`ɵɵdirectiveInject()` call, which avoids the need to special test_cases
`ChangeDetectorRef` in the generated code.

petebacondarwin commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -12,8 +12,8 @@
         "runtime-es2015": 3033,
-        "main-es2015": 452165,
-        "polyfills-es2015": 52493
+        "main-es2015": 451600,
+        "polyfills-es2015": 52215
```
Note that this PR actually decreases the payload size of angular.io, but not by as much is it appears here since `master` was already a bit lower than the previous recorded values:

file           | golden       | master        | PR            | diff (%)
----------|------------|------------|----------|-------------
main         | 452165      | 451671      | 451600     | -71 (-0.01%)
polyfills    | 52493        | 52215        | 52215       | 0 (0%)

petebacondarwin replied:
Any pipe that has non-optional injectable dependencies will have a slightly larger footprint since the `InjectFlags.ForPipe` value (`16`) must be added as an extra argument, thus adding 3 characters to the injection instruction per dependency.
It is worth noting that many pipes do not actually inject anything. A quick survey of pipes in `angular/angular` gives 7 pipes that inject dependencies:

Pipe | No. dependencies
-----|------------------
async | 1
date   | 1
i18nPlural | 1
keyvalue | 1
number | 1
percent | 1
currency | 2

petebacondarwin committed:
refactor(compiler-cli): having an `ɵinj` field no longer guarantees injectability

When `ɵngDeclareInjector()` was implemented, the `factory` was moved
out to the `ɵfac` static property on the class. This check was not updated.

AndrewKushnir commented:
[Presubmit](http://test/OCL:364369747:BASE:364369784:1616438361299:784b579c).

JoostK commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -236,23 +220,50 @@ export function compileFactoryFunction(meta: R3FactoryMetadata): R3CompiledExpre
+      `${meta.name}_Factory`);
+
+  if (baseFactoryVar !== null) {
+    // There is a base factory variable so wrap its declaration along with the factory function into
+    // an IIFE.
```
Is the IIFE really needed? Can it be emitted into the returned `statements` array instead, which would probably declare the variable after the class?

petebacondarwin replied:
That might work for full AOT but in the linker we don't have the option of adding more statements.

JoostK replied:
We may be able to emit it into the constant pool in that case. But let's land this as is right now.

petebacondarwin committed:
refactor(compiler): remove `R3ResolvedDependencyType` altogether

Now that other values were removed from `R3ResolvedDependencyType`,
its meaning can now be inferred from the other properties in the
`R3DeclareDependencyMetadata` type. This commit removes this enum
and updates the code to work without it.

JoostK commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -482,11 +482,16 @@ function convertR3DependencyMetadataArray(facades: R3DependencyMetadataFacade[]|
-    dep.attribute = new WrappedNodeExpr(attribute);
-  }
+  const wrappedAttribute = attribute === undefined ? undefined : new WrappedNodeExpr(attribute);
+  const wrappedToken = wrappedAttribute || token === null ? null : new WrappedNodeExpr(token);
```
My brain has trouble deriving the precedence here... would this be

```suggestion
  const wrappedToken = wrappedAttribute ?? (token === null ? null : new WrappedNodeExpr(token));
```

?

petebacondarwin replied:
Other refactoring has made this suggestion redundant.

JoostK commented on packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_factory_linker_1.ts:
```diff
@@ -67,15 +67,19 @@ function getDeps<TExpression>(
-  if (depObj.has('attribute')) {
-    dep.attribute = depObj.getOpaque('attribute')
-  };
+  if (attribute) {
```
```suggestion
  if (attribute !== null) {
```

But actually, `R3DependencyMetadata.attribute` should be required and nullable instead?

Thinking about it more, this is deviating from what `attribute` ought to be: a type representation of the attribute name. The linker now has to set it as a marker, but its value is not otherwise used. And the opaque `attribute` expression in the partial declaration may not be a valid type representation. For the `dynamicAttrName` case, the AOT compiler uses `token: Wrapped(Call(receiver="dynamicAttrName")), attribute: Wrapped(KeywordType(unknown))` and the linker would now provide `token: Wrapped(Call("dynamicAttrName")), attribute: Wrapped(Call(receiver="dynamicAttrName"))`. So having `R3ResolvedDependencyType` internally (but not in the partial declaration) may still be desirable.

petebacondarwin replied:
I agree that it is a bit obfuscated, but adding back in a flag is redundant when the existence of the `attribute` property is sufficient to indicate that this needs to be specially injected.

I am going to rename `attribute` to `attributeNameType` in the internal `R3DependencyMetadata` interface, since that better describes its meaning and purpose. Then in outside of full AOT (JIT mode and linking) we will pass a dummy value (the literal string `"unknown"`) into this property for attributes and I will document this in the code.

petebacondarwin replied:
Further in the `R3DeclareDependencyMetadata` we should only have the `token` and a boolean flag `attribute` since in linking we don't care about the attribute name type. This fits better with the other flags `host`, `optional`, etc.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -379,7 +379,7 @@ export interface R3DeclareDependencyMetadata {
    * not valid.
    */
-  token: o.Expression|null;
+  token?: o.Expression|null;
```
Please extend the comment to mention that either `attribute` or `token` is set, but never both or neither.

petebacondarwin replied:
Other refactoring has removed the need for this suggestion, since now there is always a `token`.

AndrewKushnir commented:
@petebacondarwin FYI, I ran presubmits and there was an error related to the `@internal` JSDoc tag use on `ForPipe`, but it looks like it's removed in the PR. Please let me know when the PR is ready for another presubmit and I'll start one. Thank you.

petebacondarwin committed:
refactor(compiler): remove unused `dependenciesFromGlobalMetadata()` function

This function is never called by any of the code in Angular and is not publicly exported.
So it can be safely removed.

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -372,19 +372,23 @@ export enum R3FactoryTarget {
+   * Default: false.
    */
-  resolved: R3ResolvedDependencyType;
+  attribute?: boolean;
```
Sorry for being a PITA, but why this change? I much prefer the way it was yesterday, where `attribute` was used instead of `token`

petebacondarwin replied:
I changed for two reasons:

- it actually made the implementation more complicated than it needed to be. In practice the `token` is always the thing that is passed to the `inject()`/`directiveInject()`/`attributeInject()` call.
- it felt magical and confusing since internally the `attribute` property (which I renamed to `attributeNameType`) was not the same as what such an `attribute` property would mean in the declaration. I believe that in the future, that would make it harder to follow what is going on.

Here an `attribute` boolean also fits with the idea that adding a decorator (e.g. `Attribute`, `Host`, `Optional` etc) causes a flag to be added to the declaration that modifies how the injection is setup.

JoostK replied:
I don't think that the implementation is more complex either way. More importantly, I see the partial declaration as representing the user's code, _not_ necessarily what is simplest for the compiler (but of course it's not as black and white as that). The user writes `@Attribute("name")` and e.g. `@Optional()`, so the partial declaration specifies the necessary information for each decorator in the property corresponding with that decorator. The attribute name is not a token, so putting it in `token` is misleading. That the compiler's metadata currently stores the attribute name in `token` shouldn't be a reason to do the same in the partial declaration; we can change the compiler metadata however we like but that's not true for the partial declaration.

So my continued preference is for the partial declaration to contain `attribute: "name"` without `token`.

petebacondarwin replied:
> The attribute name is not a token, so putting it in token is misleading

I disagree with this statement. A DI token is a thing that is used to work out what to inject in its place. In the case of an attribute injection the token is the name of the attribute.

JoostK commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -268,34 +231,35 @@ function compileInjectDependency(
+    // given for the attribute is not a string literal. For example given `@Attribute(foo())`, we
+    // want to generate `ɵɵinjectAttribute(foo())`.
+    //
+    // The `dep.attribute` is only actually used (in `createCtorDepType()`) to generate typings.
```
```suggestion
    // The `dep.attributeNameType` is only actually used (in `createCtorDepType()`) to generate typings.
```

petebacondarwin replied:
And earlier in this comment block I see!

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:364369747:BASE:364924757:1616631262696:7171cd71).

petebacondarwin committed:
fixup! refactor(core): remove the need for `ɵɵinjectPipeChangeDetectorRef()`

AndrewKushnir commented:
[Presubmit #3](http://test/OCL:364369747:BASE:365060274:1616690766021:198ab783).

JoostK commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -56,26 +54,20 @@ export interface R3ConstructorFactoryMetadata {
-   * An expression for the function which will be used to inject dependencies. The API of this
-   * function could be different, and other options control how it will be invoked.
-   */
-  injectFn: o.ExternalReference;
```
The commit message that removes this mentions `ngInject` instead of `injectFn`.

petebacondarwin replied:
Updated thanks.

JoostK commented on packages/compiler/src/render3/r3_factory.ts:
```diff
@@ -245,65 +178,89 @@ export function compileFactoryFunction(meta: R3FactoryMetadata): R3CompiledExpre
-    dep: R3DependencyMetadata, injectFn: o.ExternalReference, isPipe: boolean,
-    index: number): o.Expression {
+    dep: R3DependencyMetadata, target: R3FactoryTarget, index: number): o.Expression {
+  const injectFn = getInjectFn(target);
```
It appears this is only used in the `else if` block below.

JoostK commented on packages/core/src/compiler/compiler_facade_interface.ts:
```diff
@@ -97,20 +91,27 @@ export enum R3FactoryTarget {
 }
 
+export interface R3DeclareDependencyMetadataFacade {
+  token?: unknown;
```
This is required now

```suggestion
  token: unknown;
```

JoostK commented on packages/core/src/di/interface/injector.ts:
```diff
@@ -70,4 +70,9 @@ export const enum InternalInjectFlags {
\ No newline at end of file
+
+  /**
+   * This token is being injected into a pipe.
```
Worth mentioning here that this is intentionally not in the public-facing `InjectFlags`.

petebacondarwin replied:
Added some extra comments.

JoostK commented on packages/core/src/di/jit/util.ts:
```diff
@@ -23,24 +23,18 @@ export function reflectDependencies(type: Type<any>): R3DependencyMetadataFacade
-    meta.resolved = compiler.R3ResolvedDependencyType.Token;
-    meta.token = token;
-  }
+  } as R3DependencyMetadataFacade;
```
Was the cast intentional instead of declaring the type of `meta` as it originally was? The latter is more common.

petebacondarwin replied:
This was one of the hoops I was having to jump through when `R3DependencyMetadataFacade` was a union since it would not let me leave out both `token` and `attribute` (nor include them both). This is no longer needed since all the properties are now mandatory.

JoostK commented on packages/core/src/render3/jit/partial.ts:
```diff
@@ -31,6 +31,22 @@ export function ɵɵngDeclareComponent(decl: R3DeclareComponentFacade): unknown
+/**
+ * These enums are used in the partial factory declaration calls.
+ */
+export {R3FactoryTarget} from '../../compiler/compiler_facade';
```
Let's rename `R3FactoryTarget` to `FactoryTarget` here as well (I mean the declaration itself, not the reexport)

petebacondarwin replied:
`R3FactoryTarget` is also defined internally at https://github.com/angular/angular/blob/18b63e8d240cdef2188663f39c72d2f1ac78a163/packages/compiler/src/render3/r3_factory.ts#L81

Should we rename it there too?

JoostK replied:
We might as well, yeah

petebacondarwin replied:
I renamed across the board. There were a lot of places this enum is defined!

JoostK commented on packages/compiler/src/render3/partial/api.ts:
```diff
@@ -333,3 +333,84 @@ export interface R3DeclarePipeMetadata extends R3PartialDeclaration {
+   * If this is `'invalid'`, then one or more of the parameters wasn't resolvable and any attempt to
+   * use these deps will result in a runtime error.
+   */
+  deps: R3DeclareDependencyMetadata[]|'invalid'|null;
```
Another observation: we use `null` to indicate that the class has no constructor, which ends up generating `getInheritedFactory`. It may be worth exploring if it's possible to only generate that call if the class has an `extends` clause, sometime in the future (which requires capturing  information on whether the class has a base class in the factory declaration).

JoostK reviewed (approved):
Some more comments but nothing blocking from my side.

petebacondarwin commented on packages/compiler-cli/src/ngtsc/annotations/src/directive.ts:
```diff
@@ -946,3 +919,7 @@ const QUERY_TYPES = new Set([
 ]);
+
+function toDirectiveFactoryMetadata(meta: R3DirectiveMetadata): R3FactoryMetadata {
+  return {...meta, target: R3FactoryTarget.Directive};
```
Yeah, some slipped through in previous PRs :-)

petebacondarwin replied:
Ideally, I would like to move the factory metadata completely out from the other metadata for the injectable (e.g. directive, component, pipe, etc). But that is a bigger refactoring.

petebacondarwin committed:
refactor(compiler-cli): use a shared function for gathering factory metadata.

Each of the annotations had its own function for doing this, and those
methods were generally employing spread operators that could allow
unwanted properties to leak into the factory metadata object.

This commit supplies a shared function `toFactoryMetadata()` that
avoids this spread of properties into the returned function.

petebacondarwin commented on packages/compiler/src/jit_compiler_facade.ts:
```diff
@@ -449,31 +461,40 @@ function computeProvidedIn(providedIn: Type|string|null|undefined): Expression {
+function convertR3DeclareDependencyMetadata(facade: R3DeclareDependencyMetadataFacade):
+    R3DependencyMetadata {
+  const isAttributeDep = facade.attribute ?? false;
+  const token = facade.token == null ? null : new WrappedNodeExpr(facade.token);
```
Good spot. That is a leftover from when `token` could be optional.

petebacondarwin replied:
And because it cannot be `undefined` now we can simplify this expression further:

```suggestion
  const token = facade.token && new WrappedNodeExpr(facade.token);
```

petebacondarwin replied:
But... I know you are going to say that you prefer it to be explicit - I do - so I won't do that :-D

petebacondarwin replied:
And in fact since `token` is typed as `unknown` TS wouldn't like it either.

petebacondarwin committed:
fixup! refactor(compiler): remove `R3ResolvedDependencyType` altogether

petebacondarwin committed:
refactor(compiler): rename `R3FactoryTarget` to `FactoryTarget`

This enumeration will now start to appear in publicly facing code,
as part of declarations, so we remove the R3 to make it less specific
to the internal name for the Ivy renderer/compiler.

petebacondarwin committed:
fixup! refactor(compiler): remove `injectFn` option from factory metadata

petebacondarwin commented:
This needs public-api approval for this change https://github.com/angular/angular/pull/41231/files#diff-91d191211cf0592061a8a72cb020667a6b73e0aeea4f004fd63d4fe40e8e267e

AndrewKushnir commented:
FYI, presubmits are successful for the changes in this PR.
Note to Caretaker: there were [some failures](https://test.corp.google.com/ui#id=OCL:364369747:BASE:365111873:1616704067562:87652833), but I verified that they are pre-existing by running these targets without the changes from this PR.

AndrewKushnir reviewed (approved):
Reviewed-for: public-api
Reviewed-for: size-tracking

atscott reviewed (approved):
Reviewed-for: public-api
Reviewed-for: size-tracking

jessicajaniuk reviewed (approved):
LGTM :cookie:

reviewed-for: public-api

petebacondarwin commented:
Merge-assistance: The four outstanding reviews requests are not needed and I don't seem to be able to de-request them.
